File: packages/api/aqapi/icon_category/domain/value_object/icon_category_name.py
from dataclasses import dataclass

from aqapi.core.domain.value_object.base_value_object import BaseValueObject
from aqapi.language.domain.language_type import LanguageType
from aqapi.core.domain.base_collection import CollectionItemProtocol

@dataclass
class IconCategoryName(BaseValueObject, CollectionItemProtocol[LanguageType]):
    _language: LanguageType
    _value: str

    def __init__(self, language: LanguageType, value: str,):
        self._language = language
        super().__init__(value)

    def _validate(self) -> None:
        if not isinstance(self._value, str) or not self._value.strip():
            raise ValueError("IconCategoryNameは空文字であってはなりません。")
        
    @property
    def id(self) -> LanguageType:
        """言語タイプをIDとして返す"""
        return self._language

File: packages/backend/src/features/icon-category/value-objects/iconCategoryName.ts
import { BaseValueObject } from '@backend/core/value-object/baseValueObject';
import { LocaleString } from '@backend/core/messages/localeString';
import { LanguageTypeValue } from '@backend/features/language/value-object/languageTypeValue';
import { LanguageId } from '@backend/features/language/value-object/languageId';
import { CollectionItemProtocol } from '@backend/core/models/baseCollection';

/**
 * アイコンカテゴリ名の値オブジェクト
 */
export class IconCategoryName extends BaseValueObject<string> implements CollectionItemProtocol<LanguageId> {
  private readonly _language: LanguageTypeValue;

  constructor(language: LanguageTypeValue, value: string) {
    super({ value });
    this._language = language;
  }

  /**
   * 言語タイプをIDとして返す
   */
  get id(): LanguageId {
    return this._language.id;
  }

  /**
   * 言語タイプを取得
   */
  get language(): LanguageTypeValue {
    return this._language;
  }

  /**
   * バリデーション実行
   */
  protected validate(): void {
    this.validator.required();
  }

  /**
   * 値オブジェクトの名前を取得
   */
  protected get valueName(): LocaleString {
    return new LocaleString({
      ja: 'アイコンカテゴリ名',
      en: 'Icon Category Name'
    });
  }
}
